### ISSUE TITLE
Linq performance issue with non cacheable linq query plan 

### ISSUE THREAD

**Issue by gokhanabatay** [nhibernate/nhibernate-core#2952]
Hi,
We have a performance issue with below linq query is not added to query plan cache. Because x.F2.Mask(6, 4, '*', MaskOption.Default) of this function call makes query uncacheable. This function is masks card number middle digits other than first 6 and last 4 digits. This api called concurrent 100 request at the same time.

First of all why this function call leads to query plan uncacheable, is this desired behaviour?
When we look at Dynatrace pure path below image added there is a suspension at QueryTranslatorImpl.Compile that I cannot understand and takes to much time to compile the query?

To report all uncacheable queries could you change QueryPlanCache.GetHqlQueryPlan log.Debug("Query plan not cacheable") to log.Warn("Query plan not cacheable {0}", queryExpression.Key) Adding elapsed time could be better for performance issues, we cannot find this issue without Dynatrace.

Environment
NHibernate 5.3.9
.net 5.0
linux container

```
            var transactions = Query<TxnAcquirerOnl>().WithOptions(opt => opt.SetReadOnly(true))
                                                   .Where(predicateTxnAcquirerOnl)
                                                   .OrderByDescending(order => order.Guid)
                                                   .Select(x => new GetTransactionsResponse
                                                   {
                                                       TxnGuid = x.Guid,
                                                       CardNo = x.F2.Mask(6, 4, '*', MaskOption.Default),
                                                       CardDci = x.CardDci,
                                                       CardSource = x.CardSource,
                                                       TerminalType = x.TerminalType,
                                                       TransactionName = x.TxnDef.Description.GetText(CurrentLanguage),
                                                       TxnEffect = x.TxnDef.TxnEffect,
                                                       TxnEntry = x.TxnEntry,
                                                       TransactionStatus = x.TxnStt,
                                                       TransactionDate = x.F13,
                                                       TransactionTime = x.F12,
                                                       TxnRegion = x.TxnRegion,
                                                       BinProductType = x.BinProductType,
                                                       MerchantCode = x.F42,
                                                       MerchantName = x.F43Name,
                                                       MerchantLocation = x.F43,
                                                       MerchantCity = x.F43City,
                                                       MerchantCountry = x.F43Country,
                                                       TerminalCode = x.F41,
                                                       TerminalStan = x.F11Trm,
                                                       BatchNo = x.BatchNo,
                                                       InstallCount = x.TxnAcquirerInstall.InstallCnt,
                                                       AuthCode = x.F38,
                                                       Rrn = x.F37,
                                                       ResponseCode = x.F39,
                                                       ResponseDescription = x.TxnResponseCodeDef.Description.GetText(CurrentLanguage),
                                                       Irc = x.Irc,
                                                       IrcDescription = x.TxnIrcDef.Description.GetText(CurrentLanguage),
                                                       Mcc = x.F18,
                                                       RequestDate = x.RequestDate,
                                                       RequestTime = x.RequestTime,
                                                       TransactionAmount = x.F4,
                                                       TransactionCurrency = x.F49,
                                                       BillingAmount = x.F6,
                                                       BillingCurrency = x.F51,
                                                   }).ToList()
```

SQL:
```
select
	txnacquire0_.guid as col_0_0_,
	txnacquire0_.f2 as col_1_0_,
	txnacquire0_.card_dci as col_2_0_,
	txnacquire0_.card_source as col_3_0_,
	txnacquire0_.terminal_type as col_4_0_,
	OCN_CFG.FN_GET_LANGUAGE_TEXT(txndef1_.description,
	:p0) as col_5_0_,
	txndef1_.txn_effect as col_6_0_,
	txnacquire0_.txn_entry as col_7_0_,
	txnacquire0_.txn_stt as col_8_0_,
	txnacquire0_.f13 as col_9_0_,
	txnacquire0_.f12 as col_10_0_,
	txnacquire0_.txn_region as col_11_0_,
	txnacquire0_.bin_product_type as col_12_0_,
	txnacquire0_.f42 as col_13_0_,
	txnacquire0_.f43_name as col_14_0_,
	txnacquire0_.f43 as col_15_0_,
	txnacquire0_.f43_city as col_16_0_,
	txnacquire0_.f43_country as col_17_0_,
	txnacquire0_.f41 as col_18_0_,
	txnacquire0_.f11_trm as col_19_0_,
	txnacquire0_.batch_no as col_20_0_,
	txnacquire2_.install_cnt as col_21_0_,
	txnacquire0_.f38 as col_22_0_,
	txnacquire0_.f37 as col_23_0_,
	txnacquire0_.f39 as col_24_0_,
	OCN_CFG.FN_GET_LANGUAGE_TEXT(txnrespons3_.description,
	:p1) as col_25_0_,
	txnacquire0_.irc as col_26_0_,
	OCN_CFG.FN_GET_LANGUAGE_TEXT(txnircdef4_.description,
	:p2) as col_27_0_,
	txnacquire0_.f18 as col_28_0_,
	txnacquire0_.request_date as col_29_0_,
	txnacquire0_.request_time as col_30_0_,
	txnacquire0_.f4 as col_31_0_,
	txnacquire0_.f49 as col_32_0_,
	txnacquire0_.f6 as col_33_0_,
	txnacquire0_.f51 as col_34_0_
from
	OCN_ACQ.TXN_ACQUIRER_ONL txnacquire0_
left outer join OCN_CFG.TXN_DEF txndef1_ on
	txnacquire0_.txn_def_guid = txndef1_.guid
left outer join OCN_ACQ.TXN_ACQUIRER_INSTALL txnacquire2_ on
	txnacquire0_.guid = txnacquire2_.txn_guid
left outer join OCN_CFG.TXN_RESPONSE_CODE_DEF txnrespons3_ on
	txnacquire0_.f39 = txnrespons3_.code
left outer join OCN_CFG.TXN_IRC_DEF txnircdef4_ on
	txnacquire0_.irc = txnircdef4_.code
where
	txnacquire0_.STATUS = :p3
	and txnacquire0_.MBR_ID = :p4
	and txnacquire0_.f42 =:p5
	and cast(cast(txnacquire0_.f13 as text)|| LPAD(cast(txnacquire0_.f12 as text), '*****', '*****') as numeric('*****'))<=:p6
	and cast(cast(txnacquire0_.f13 as text)|| LPAD(cast(txnacquire0_.f12 as text), '*****', '*****') as numeric('*****'))>=:p7
order by
	txnacquire0_.guid desc
```

Suspension: total 51.5 ms is too much
![image](https://user-images.githubusercontent.com/9988911/144202806-f1565379-53cd-492e-ba4d-632f236006a7.png)

Another pure path without Suspension but 1.72 ms too much for this kind of query for every request
![image](https://user-images.githubusercontent.com/9988911/144206046-f0a9c1d2-fc11-4702-8302-6ffa8681b5ae.png)

**Comment by razzmatazz** [comment]
@gokhanabatay unrelated, but what is the tool you are using for profiling (in screenshots)?

**Comment by bahusoid** [comment]
> First of all why this function call leads to query plan uncacheable, is this desired behaviour?

Well some parametrized lambda is executed in code rather than SQL. And those constant parameters are not part of query plan - such caching is not supported (see https://github.com/nhibernate/nhibernate-core/pull/2375 which might fix it)

If you have any methods which are not translated to SQL directly - try to move them out of query. So my suspect is ` CardNo = x.F2.Mask(6, 4, '*', MaskOption.Default)`. Currently it would be more effective to do it in separate cycle after query execution.

**Comment by gokhanabatay** [comment]
> @gokhanabatay unrelated, but what is the tool you are using for profiling (in screenshots)?

We are using [Dynatrace](https://www.dynatrace.com/)

**Comment by gokhanabatay** [comment]
> > First of all why this function call leads to query plan uncacheable, is this desired behaviour?
> 
> Well some parametrized lambda is executed in code rather than SQL. And those constant parameters are not part of query plan - such caching is not supported (see #2375 which might fix it)
> 
> If you have any methods which are not translated to SQL directly - try to move them out of query. So my suspect is ` CardNo = x.F2.Mask(6, 4, '*', MaskOption.Default)`. Currently it would be more effective to do it in separate cycle after query execution.

When we remove Mask query plan is cached but suspencion is the main problem for queries that is not cacheable(first screenshot), I could not find out reason of suspencion? Dou you have any idea? Query compile is too much expensive 51.5 ms #2375 when will you deploy this change.

> To report all uncacheable queries could you change QueryPlanCache.GetHqlQueryPlan log.Debug("Query plan not cacheable") to log.Warn("Query plan not cacheable {0}", queryExpression.Key) Adding elapsed time could be better for performance issues, we cannot find this issue without Dynatrace.

Also what dou you think about logging suggestion?

**Comment by bahusoid** [comment]
> but suspencion is the main problem for queries that is not cacheable

No idea about suspension. Again the only suggestion I have: just make problematic queries cacheable again by  executing in code methods after query execution. 

>  #2375 when will you deploy

When it's ready and properly reviewed (not me).

> Also what do you think about logging suggestion

No objections

**Comment by gokhanabatay** [comment]
@bahusoid 
When i go deep down I see an issue with all queries except `Query 5` at `DefaultQueryProvider.ExecuteList` self time is too much.
`Query 5` difference is there is no `Mask` method call with `LinqExtensionMethodAttribute`.

Query 1:
![image](https://user-images.githubusercontent.com/9988911/144514134-e43673f7-3c48-4766-86e3-c3f7e623b994.png)

Query 2:
![image](https://user-images.githubusercontent.com/9988911/144514643-3081aeec-fd26-41a9-ae0c-6ec122767678.png)

Query 3:
![image](https://user-images.githubusercontent.com/9988911/144514705-f646da47-c087-4589-8aa0-fa68829b96b7.png)

Query 4:
![image](https://user-images.githubusercontent.com/9988911/144514801-541824a0-0b88-49b3-8752-1b3ee1e3515b.png)

Query 5:
![image](https://user-images.githubusercontent.com/9988911/144514929-4fa05cbe-663c-4313-bfc4-33ae381586d8.png)

With 'Query 5' there is 893 micro seconds self time at `DefaultQueryProvider.ExecuteList` its acceptable so I think there is a perfromance issue with `LinqExtensionMethodAttribute`

Our mask method calls database function in sql and registered through session factory setup.
```
[LinqExtensionMethod]
 public static string Mask(this string value, int firstCharacters, int lastCharacters, char mask, MaskOption maskOption = MaskOption.Default)
{
     throw new System.NotSupportedException();
}

public enum MaskOption
{
        /// <summary>
        /// if (first + last characters).length = input.length
        /// Returns input value
        /// </summary>
        Default,
        /// <summary>
        /// if (first + last characters).length = input.length
        /// Prefer first characters to be open
        /// </summary>
        PreferFirstCharacterToBeOpen,
        /// <summary>
        /// if (first + last characters).length = input.length
        /// Prefer last characters to be open
        /// </summary>
        PreferLastCharacterToBeOpen
}
```
 
So I think the issue is in `NhPartialEvaluatingExpressionVisitor` but I cannot figure out which one, how to find out reason can you help?

```
public static PreTransformationResult PreTransform(Expression expression, PreTransformationParameters parameters)
{
		parameters.EvaluatableExpressionFilter = new NhEvaluatableExpressionFilter(parameters.SessionFactory);
		parameters.QueryVariables = new Dictionary<ConstantExpression, QueryVariable>();

		var partiallyEvaluatedExpression = NhPartialEvaluatingExpressionVisitor
				.EvaluateIndependentSubtrees(expression, parameters);

		return new PreTransformationResult(
				parameters.PreTransformer.Invoke(partiallyEvaluatedExpression),
				parameters.SessionFactory,
				parameters.QueryVariables);
}
```

**Comment by gokhanabatay** [comment]
We changed nuget to `NHibernate 5.4.0-dev.3582` and I guess find the issue @bahusoid @razzmatazz, that is similar you resolved with #2948

`ExpressionProcessor` compile call can we change it, its is really effects the performance very bad you can look below dynatrace outputs.
```
public static object FindValue(Expression expression)
{
	if (expression.NodeType == ExpressionType.Constant)
		return ((ConstantExpression) expression).Value;

	if (expression.NodeType == ExpressionType.MemberAccess)
	{
		var memberAccess = (MemberExpression) expression;
		if (memberAccess.Expression == null || memberAccess.Expression.NodeType == ExpressionType.Constant)
		{
			var constantValue = ((ConstantExpression) memberAccess.Expression)?.Value;
			var member = memberAccess.Member;
			switch (member.MemberType)
			{
				case MemberTypes.Field:
					return ((FieldInfo) member).GetValue(constantValue);
				case MemberTypes.Property:
					return ((PropertyInfo) member).GetValue(constantValue);
			}
		}
	}

	var valueExpression = Expression.Lambda(expression).Compile();
	object value = valueExpression.DynamicInvoke();
	return value;
}
```
![image](https://user-images.githubusercontent.com/9988911/144603777-0b41b85f-d0c1-4e27-b816-3011f42fc833.png)

![image](https://user-images.githubusercontent.com/9988911/144603835-63a20a98-1499-4521-a052-666dbdd38fc2.png)

![image](https://user-images.githubusercontent.com/9988911/144602894-377b8fb6-7372-4333-b156-a66100740a0b.png)

**Comment by bahusoid** [comment]
>  I guess find the issue 

Are you saying the perf is worse than in 5.3 or what? All your images are not helpful.

Compilation is not an issue. Though it can be avoided. Find what expressions are compiled and move them out of query in to parameter variables. Example:

```C#
int GetValue(int p1) => p1 *1000;
query.Where(e => e.Id == GetValue(1)); <-- GetValue(1) call will be compiled
```

With #2948 you can avoid compilation by moving GetValue method call out of query:
```C#
int v = GetValue(1);
query.Where(e => e.Id == v); <-- No compilation
```

And as I already explained in #2947 if  `e.Id` from example above is nullable (`int?`) or `long` it will still compile. `v` and `e.Id` must be exactly of the same types to avoid compilation.

**Comment by gokhanabatay** [comment]
@bahusoid  performance is better than 5.3 thanks for that, let me try to tell this image.

ExpressionProssor.FindValue called 6 times and it took 5.49 ms. 
![image](https://user-images.githubusercontent.com/9988911/144619968-4f3cc8b6-ae95-4436-94d5-e230aa5d159c.png)

2xToLongDateTime() returns "YYYYMMDDHHMMSS" as long , I think this extension method compile and dynamic invoke takes times.
I added debug view one of calls
![image](https://user-images.githubusercontent.com/9988911/144619812-40dd4ce7-c0dd-48e3-8e45-11fa9a83d8b8.png)

3xCurrentLanguage="en-US" constant so its ok
1xMaskOption.Default = 0 constant so its ok

ConcatToLongDateTime, GetText and Mask methods are `LinqExtensionMethodAttribute` so they are not executed.

```c#
 var predicateTxnInfoAcquirer = PredicateBuilder.True<TxnInfoAcquirer>()
               .And(x => x.F42 == request.MerchantId)
               .And(x => request.MerchantIds.Contains(x.F42))
               .And(x => x.F41 == request.TerminalId)
               .And(x => request.TerminalIds.Contains(x.F41))
               .And(x => x.F13.Value.ConcatToLongDateTime(x.F12.Value) <= request.EndDate.ToLongDateTime())
               .And(x => x.F13.Value.ConcatToLongDateTime(x.F12.Value) >= request.StartDate.ToLongDateTime());

var txnAcquirer = Query<TxnAcquirer>().WithOptions(opt => opt.SetReadOnly(true))
                                      .Where(predicateTxnAcquirer)
                                      .OrderByDescending(order => order.Guid)
                                      .Select(x => new GetTransactionsResponse
                                      {
                                          TxnGuid = x.Guid,
                                          CardNo = x.F2.Mask(6, 4, '*', MaskOption.Default),
                                          CardDci = x.CardDci,
                                          CardSource = x.CardSource,
                                          TerminalType = x.TerminalType,
                                          TransactionName = x.TxnDef.Description.GetText(CurrentLanguage),
                                          TxnEffect = x.TxnDef.TxnEffect,
                                          TxnEntry = x.TxnEntry,
                                          TransactionStatus = x.TxnStt,
                                          TransactionDate = x.F13,
                                          TransactionTime = x.F12,
                                          TxnRegion = x.TxnRegion,
                                          BinProductType = x.BinProductType,
                                          MerchantCode = x.F42,
                                          MerchantName = x.F43Name,
                                          MerchantLocation = x.F43,
                                          MerchantCity = x.F43City,
                                          MerchantCountry = x.F43Country,
                                          TerminalCode = x.F41,
                                          TerminalStan = x.F11Trm,
                                          BatchNo = x.BatchNo,
                                          InstallCount = x.TxnAcquirerInstall.InstallCnt,
                                          AuthCode = x.F38,
                                          Rrn = x.F37,
                                          ResponseCode = x.F39,
                                          ResponseDescription = x.TxnResponseCodeDef.Description.GetText(CurrentLanguage),
                                          Irc = x.Irc,
                                          IrcDescription = x.TxnIrcDef.Description.GetText(CurrentLanguage),
                                          Mcc = x.F18,
                                          RequestDate = x.RequestDate,
                                          RequestTime = x.RequestTime,
                                          TransactionAmount = x.F4,
                                          TransactionCurrency = x.F49,
                                          BillingAmount = x.F6,
                                          BillingCurrency = x.F51,
                                      }).ToList();
```

If we change expression `Expression.Lambda(expression).Compile().DynamicInvoke()`  to expession.method.invoke(***) something like this, does it resolve the performance issue?

**Comment by gokhanabatay** [comment]
Another query getting user roles causes expression compile `key.Id` just member access
if you guide me i can do the development.

```c#
         claims = Query<EntUserRoleOwnership>()
                     .Where(x => x.UserCode == key.Id)
                     .Select(x => x.RoleGuid)
                     .WithOptions(options =>
                     {
                            options.SetReadOnly(true);
                     })
                    .ToList();
```

![image](https://user-images.githubusercontent.com/9988911/144664383-763c7b05-afd2-4be8-bc84-d527fc9092aa.png)

**Comment by gokhanabatay** [comment]
https://github.com/nhibernate/nhibernate-core/issues/2952#issuecomment-985076657 fixed #2957, throughput increased %100 from 2K tps to 4K tps
